components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+			};
 		}
+		return null;
 	}
 
-	search( event ) {
-		const { triggerPrefix } = this.props;
-		const { isOpen } = this.state;
-
-		const value = event.target.textContent;
+	// this method is separate so it can be overrided in tests
+	createRange( startNode, startOffset, endNode, endOffset ) {
+		const range = document.createRange();
+		range.setStart( startNode, startOffset );
+		range.setEnd( endNode, endOffset );
+		return range;
+	}
 
-		// Currently search supports only if input's entire value is match
-		let pattern = '^';
+	loadOptions( index ) {
+		this.props.completers[ index ].getOptions().then( ( options ) => {
+			this.setState( {
+				[ 'options_' + index ]: options.map(
+					( option, i ) => ( { ...option, key: index + '_' + i } ) ),
+			} );
+		} );
+	}
 
-		if ( triggerPrefix ) {
-			// Trigger prefix may contain reserved characters of a regular
-			// expression, so escape
-			pattern += escapeStringRegexp( triggerPrefix );
+	findMatch( container, cursor, allCompleters, wasOpen ) {
+		const allowAnything = () => true;
+		let endTextNode;
+		let endIndex;
+		let completers = allCompleters.map( ( completer, idx ) => ( { ...completer, idx } ) );
+		if ( wasOpen ) {
+			// put the open completer at the start so it has priority
+			completers = [
+				wasOpen,
+				...completers.filter( ( completer ) => completer.idx !== wasOpen.idx ),
+			];
 		}
-
-		// Create matching group for the search value
-		pattern += '([^\\s]*)$';
-
-		pattern = new RegExp( pattern );
-
-		let match, search;
-		if ( ( match = value.match( pattern ) ) ) {
-			search = new RegExp( escapeStringRegexp( match[ 1 ] ), 'i' );
+		// search backwards to find the first preceeding space or non-text node.
+		if ( isTextNode( cursor.node ) ) { // TEXT node
+			endTextNode = cursor.node;
+			endIndex = cursor.offset;
+		} else if ( cursor.offset === 0 ) {
+			endTextNode = onlyTextNode( descendFirst( cursor.node ) );
+			endIndex = 0;
+		} else {
+			endTextNode = onlyTextNode( descendLast( cursor.node.childNodes[ cursor.offset - 1 ] ) );
+			endIndex = endTextNode ? endTextNode.nodeValue.length : 0;
 		}
-
-		if ( isOpen ) {
-			if ( match ) {
-				// Reset selection to initial when search changes
-				this.setState( {
-					selectedIndex: 0,
-					search,
-				} );
+		if ( endTextNode === null ) {
+			return null;
+		}
+		// filter the completers to those that could handle this node
+		completers = completers.filter(
+			( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+		// exit early if nothing can handle it
+		if ( completers.length === 0 ) {
+			return null;
+		}
+		let startTextNode = endTextNode;
+		let text = endTextNode.nodeValue.substring( 0, endIndex );
+		let pos = lastIndexOfSpace( text );
+		while ( pos === -1 ) {
+			const prev = onlyTextNode( startTextNode.previousSibling );
+			if ( prev ) {
+				// filter the completers to those that could handle this node
+				completers = completers.filter(
```

Comment:
In this performance-critical code, I might suggest opting for Lodash equivalents to native functions, which tend to perform better:

https://jsperf.com/lodash-vs-underscore-vs-native-filter/2

Reply:
Well lodash does seem to get more stable performance across browsers but it is only the best on Chrome. Still considering we are talking about filtering a list of 3 items worrying about this is kinda pointless.

I performed these tests on the same MacBook Air for consistency.

## Firefox

Filter | OPS
------ | ---
Native filter | 92,841 ±6.30% fastest
Lodash filter | 47,316 ±2.99% 47% slower
Underscore filter | 38,479 ±8.49% slower

## Chrome

Filter | OPS
------ | ---
Native filter | 11,803 ±2.06% 76% slower
Lodash filter | 49,174 ±1.89% fastest
Underscore filter | 35,244 ±1.94% slower

## Safari

Filter | OPS
------ | ---
Native filter | 22,662 ±9.40% 64% slower
Lodash filter | 43,801 ±12.18% 32% slower
Underscore filter | 61,778 ±7.03% fastest

Reply:
Note I have changed everything to use lodash's filter and map.